### PR TITLE
Multi-line timeout

### LINKED ISSUE #140: implement multi-line-timeout
well, it seems that syslog-ng does not detect any files in this mode and/or at least not displaying any incoming message.
To reproduce the issue: create the source file with this content:
```
root@thor-t410:/home/pzolee/zwa/projects/syslog-ng-ose-3.6/install# cat /var/testdb_working_dir/073358cb-d1ab-415a-a2a3-a874a500b080.txt
Mar 18 12:07:04 testhost localprg PREFIX msg_with_prefix_without_garbage first line
msg_with_prefix_without_garbage second line
msg_with_prefix_without_garbage third line
```

then start syslog-ng with the config below.
The result:
```
[2014-06-18T09:43:28.842848] Module loaded and initialized successfully; module='syslogformat'
[2014-06-18T09:43:28.843001] Running application hooks; hook='1'
[2014-06-18T09:43:28.843019] Running application hooks; hook='3'
[2014-06-18T09:43:28.843054] syslog-ng starting up; version='3.6.0alpha0'
```
Thats all, syslog-ng does not recognize the file.

Expected result: syslog-ng processes the file then creates the proper multi line message when the timeout is reached:
"A message is considered complete if no new lines arrive to the message for 10 seconds, even if no line matching the multi-line-garbage() option is received."

## config
```
@version: 3.6
options {
    threaded(yes);
};
source s_file_97358a54e6924af4826d9ab6ccb074fb {
    file("/var/testdb_working_dir/073358cb-d1ab-415a-a2a3-a874a500b080.txt" keep_hostname(yes)
 multi-line-mode(regexp)
 multi_line_prefix("PREFIX")
);
};
destination d_file_3f633b65908846c085561437c13d2ff8 {
    file("/var/testdb_working_dir/d333cb26-8a64-4bc1-8089-b04c1ace8a20.txt");
};
log {
source(s_file_97358a54e6924af4826d9ab6ccb074fb);
destination(d_file_3f633b65908846c085561437c13d2ff8);

flags(flow-control);
};
```

### PR DESCRIPTION
**Pull request by MrAnno**
This pull request adds a new option to multi-line file sources: `multi-line-timeout()`.

After waiting `multi-line-timeout()` seconds without reading new data from the file, the last (potentially partial) message will be flushed and sent through the pipeline as a `LogMessage`.

Since the multi-line file source detects the end of a message after finding the beginning of the subsequent message (indented or no-garbage/suffix mode), this option can be used to flush the last multi-line message in the file after a `multi-line-timeout()`-second timeout.

TODO:
- [x] fix a task ordering issue
- [x] enable internal tests (prefix-nogarbage/suffix, indented mode)
- [x] check encoding (raw buffer positions)
- [x] investigate and fix a crash (when writing to the file at the time of the timeout, the flush I/O job is somehow triggered by both the timeout and the poller logic at the same time)

Buffers should also be flushed when the file is moved/restarted. It requires additional work, because `reopen()` should be scheduled after the async flush operation; this will be added later.

Resolves #140

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng retest this please

**Comment by MrAnno** [comment]
@kira-syslogng do perftest

**Comment by kira-syslogng** [comment]
Kira-performance-test: Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng do perftest

**Comment by kira-syslogng** [comment]
Kira-performance-test: Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng retest this please
@kira-syslogng do perftest

**Comment by MrAnno** [comment]
@kira-syslogng do perftest

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Kira-performance-test: Build SUCCESS

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
We've discussed IRL that I should set the default timeout to 10 seconds in case of non-garbage/non-suffix multi line sources. I'll do that soon and remove the WIP flag.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
Looks good to me. I would have approved if the PR would not be WIP and there would not be an ordering issue mentioned in the todo part of the description (which I could not spot during the review. Is it tricky?).

**Comment by MrAnno** [comment]
Thanks for the review.

It's because I use ivykis timers and there is no guaranteed ordering between 2 timers (`follow-freq()` and `multi-line-timeout()`). This could cause unexpected flush calls. I'm replacing the timer with a `time_t` field and a `follow-freq()`-based check (similar to what we have in PE 6).

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
I've resolved the crash caused by a race condition, hence removing the WIP flag.

Almost every CI quality check fails though (func-test, etc.), I'm working on it.

In the mean time, please review the code.

**Comment by gaborznagy** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
@kira-syslogng test this please test branch=multi-line-timeout;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@gaborznagy Thanks for fixing and finishing this! :+1:

File: modules/affile/poll-multiline-file-changes.c [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  poll_multiline_file_changes_stop_timer(self);
+  self->timed_out = FALSE;
+}
+
+static void
+poll_multiline_file_changes_update_watches(PollEvents *s, GIOCondition cond)
+{
+  PollMultilineFileChanges *self = (PollMultilineFileChanges *) s;
+
+  if (!cond)
```
[the remark this answers is not in the document]
Comment (gaborznagy) [reply modules/affile/poll-multiline-file-changes.c]: We discussed this IRL, and see a possibility to refactor some things, but it's not scope of this PR.

**Pull request merged**